akka-persistence-typed/src/test/scala/docs/akka/persistence/typed/AccountExample5b.scala

```diff
@@ -0,0 +1,121 @@
+    case object CreateAccount extends AccountCommand
+    case class Deposit(amount: Double) extends AccountCommand
+    case class Withdraw(amount: Double) extends AccountCommand
+    case object CloseAccount extends AccountCommand
+
+    sealed trait AccountEvent
+    case object AccountCreated extends AccountEvent
+    case class Deposited(amount: Double) extends AccountEvent
+    case class Withdrawn(amount: Double) extends AccountEvent
+    case object AccountClosed extends AccountEvent
+
+    sealed trait Account {
+      def applyEvent(event: AccountEvent): Account
+    }
+    case object EmptyAccount extends Account {
+      override def applyEvent(event: AccountEvent): Account = event match {
+        case AccountCreated ⇒ OpenedAccount(0.0)
+        case _              ⇒ throw new IllegalStateException(s"unexpected event [$event] in state [EmptyAccount]")
+      }
+    }
+    case class OpenedAccount(balance: Double) extends Account {
+      override def applyEvent(event: AccountEvent): Account = event match {
+        case Deposited(amount) ⇒ copy(balance = balance + amount)
+        case Withdrawn(amount) ⇒ copy(balance = balance - amount)
+        case AccountClosed     ⇒ ClosedAccount
+        case _                 ⇒ throw new IllegalStateException(s"unexpected event [$event] in state [OpenedAccount]")
+      }
+    }
+    case object ClosedAccount extends Account {
+      override def applyEvent(event: AccountEvent): Account =
+        throw new IllegalStateException(s"unexpected event [$event] in state [ClosedAccount]")
+    }
+
+    private def initialHandler(cmd: AccountCommand): Effect[AccountEvent, Account] =
+      cmd match {
+        case CreateAccount ⇒ Effect.persist(AccountCreated)
+        case _             ⇒ Effect.unhandled
+      }
+
+    private def openedAccountHandler(acc: OpenedAccount, cmd: AccountCommand): Effect[AccountEvent, Account] = {
+      cmd match {
+        case Deposit(amount) ⇒ Effect.persist(Deposited(amount))
+
+        case Withdraw(amount) ⇒
+          if ((acc.balance - amount) < 0.0)
+            Effect.unhandled // TODO replies are missing in this example
+          else {
+            Effect
+              .persist(Withdrawn(amount))
+              .thenRun {
+                case OpenedAccount(balance) ⇒
+                  // do some side-effect using balance
+                  println(balance)
+                case _ ⇒ throw new IllegalStateException
+              }
+          }
+        case CloseAccount if acc.balance == 0.0 ⇒
+          Effect.persist(AccountClosed)
+
+        case CloseAccount ⇒
+          Effect.unhandled
+      }
+    }
+
+    private def closedHandler(cmd: AccountCommand): Effect[AccountEvent, Account] =
+      Effect.unhandled
+
+    // slightly more verbose than the CommandHandler type alias, but it's more clear that it's just a function
+
+    private val commandHandler: (Account, AccountCommand) ⇒ Effect[AccountEvent, Account] = {
```

Comment:
+1 to this being explicit rather than the `byState`. Also if users want to encode initialState as an Option it would only appear in this match. All the command handlers that need a present state wouldn't know about the option.